sqlite/sqlite.go

```diff
@@ -74,6 +104,163 @@ func (s *SqlStore) Flush(ctx context.Context) {
 	s.log.Debug("sqlite data flushed successfully")
 }
 
+// BackupSqlStore creates a new temporary database and uses the sqlite backup API
+// to back the database up into the temporary database. It then writes the temporary
+// database file to the writer. Using the sqlite backup API allows the backup to be
+// performed without needing to lock the database, and also allows it to work with
+// in-memory databases. See: https://www.sqlite.org/backup.html
+//
+// The backup works by copying the SOURCE database to the DESTINATION database.
+// The SOURCE is the running database that needs to be backed up, and the DESTINATION
+// is the resulting backup. The underlying sqlite connection is needed for both
+// SOURCE and DESTINATION databases to use the sqlite backup API made available by the
+// go-sqlite3 driver.
+func (s *SqlStore) BackupSqlStore(ctx context.Context, w io.Writer) error {
+	span, _ := tracing.StartSpanFromContext(ctx)
+	defer span.Finish()
+
+	// create a destination db in a temporary directory to hold the backup.
+	tempDir, err := ioutil.TempDir("", "")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tempDir)
+
+	destPath := tempDir + "/" + DefaultFilename
```

Comment:
I'm not sure if this will cause problems on Windows, could we use `filepath.Join` instead?

Reply:
Good catch, made the change!